framework/src/interfaces/FunctorInterface.C

```diff
@@ -0,0 +1,96 @@
+//* This file is part of the MOOSE framework
+//* https://www.mooseframework.org
+//*
+//* All rights reserved, see COPYRIGHT for full restrictions
+//* https://github.com/idaholab/moose/blob/master/COPYRIGHT
+//*
+//* Licensed under LGPL 2.1, please see LICENSE for details
+//* https://www.gnu.org/licenses/lgpl-2.1.html
+
+#include "FunctorInterface.h"
+
+InputParameters
+FunctorInterface::validParams()
+{
+  return emptyInputParameters();
+}
+
+FunctorInterface::FunctorInterface(const MooseObject * const moose_object)
+  : _fi_params(moose_object->parameters()),
+    _fi_name(_fi_params.get<std::string>("_object_name")),
+    _fi_subproblem(*_fi_params.getCheckedPointerParam<SubProblem *>("_subproblem")),
+    _fi_tid(_fi_params.get<THREAD_ID>("_tid"))
+{
+}
+
+std::string
+FunctorInterface::deduceFunctorName(const std::string & name) const
+{
+  if (_fi_params.have_parameter<MooseFunctorName>(name))
+    return _fi_params.get<MooseFunctorName>(name);
+  // variables, functor material properties, and functions are also functors
+  else if (_fi_params.have_parameter<MaterialPropertyName>(name))
+    return _fi_params.get<MaterialPropertyName>(name);
+  else if (_fi_params.have_parameter<VariableName>(name))
+    return _fi_params.get<VariableName>(name);
+  else if (_fi_params.have_parameter<std::vector<VariableName>>(name))
+  {
+    const auto & var_names = _fi_params.get<std::vector<VariableName>>(name);
+    if (var_names.size() != 1)
+      mooseError("We only support a single variable name for retrieving a functor");
+    return var_names[0];
+  }
+  else if (_fi_params.have_parameter<NonlinearVariableName>(name))
+    return _fi_params.get<NonlinearVariableName>(name);
+  else if (_fi_params.have_parameter<FunctionName>(name))
+    return _fi_params.get<FunctionName>(name);
+  else
+    return name;
```

Comment:
You think a user might do this?
```
   functor = 'var'
   functor = 'prop'
```

Reply:
no I think a user will define both a 'rho' matprop and 'rho' variable and expect variable to be retrieved

Though is it? because the MooseFunctor is retrieved first

Reply:
We now have error checking for this. If:
- An input file yields two different functors of the same name
- `getFunctor` is hit at run-time

then you will get an error saying that the request is ambiguous